enterprise/backend/src/metabase_enterprise/transforms/util.clj

```diff
@@ -433,6 +433,85 @@
       (some->> (:router-database-id db-or-id)
                (t2/exists? :model/DatabaseRouter :database_id))))
 
+;;; ------------------------------------------------- Source Table Resolution -----------------------------------------
+
+(defn batch-lookup-table-ids
+  "Batch lookup table IDs from ref maps. Returns {[db_id schema name] -> table_id}.
+  Queries the exact conjunction of each [database_id, schema, table] triple rather than
+  a Cartesian product of all values."
+  [refs]
+  (when (seq refs)
+    (let [unique-refs (distinct (map (juxt :database_id :schema :table) refs))
+          ref->clause (fn [[db-id schema table-name]]
+                        [:and
+                         [:= :db_id db-id]
+                         (if (some? schema)
+                           [:= :schema schema]
+                           [:is :schema nil])
+                         [:= :name table-name]])]
+      (t2/select-fn->fn (juxt :db_id :schema :name) :id [:model/Table :id :db_id :schema :name]
```

Comment:
I'm a bit nervous about this exploding when there are an inordinate number of refs.

Transforms probably aren't super high cardinality, but partition-all could be good insurance.